server/src/platform/drbg/mod.rs

```diff
@@ -0,0 +1,128 @@
+//! Deterministic Random Bit Generator. Heavily inspired by
+//! <https://docs.rs/rand_core>
+
+use core::fmt::Debug;
+pub mod helpers;
+
+/// This trait wraps functionalities common to all Deterministic Random Bit
+/// Generators.
+///
+/// Three different methods for generating random data are required similar to
+/// [RngCore](https://docs.rs/rand_core/latest/rand_core/trait.RngCore.html).
+/// There is no required relationship between the output of each; e.g. an
+/// implementations of [`fill_bytes`] may consume a whole number of u32 or
+/// u64 values and drop any remaining unused bytes. The same can happen
+/// with the [`next_u32`] and [`next_u64`] methods, implementations may
+/// discard some random bits for efficiency.
+/// Implementation should normally have portable, reproducible output,
+/// i.e. fix Endianness when converting values to avoid platform
+/// differences, and avoid making any changes which affect output.
+///
+/// A DRBG may implement only one of the [`next_u32`], [`next_u64`],
+/// [`fill_bytes`] methods available in this trait directly, then use
+/// the helper functions from the [helpers] module to implement the
+/// other methods.
+///
+/// ## Implementing [`Drbg`] for large entropy or nonce
+///
+/// Note that the required traits [`core::default::Default`] and
+/// [`core::convert::AsMut<u8>`] are not implemented for large arrays
+/// [u8; N] with N > 32. To be able to implement the traits required
+/// by SeedableRng for RNGs with such large seeds, the newtype pattern
+/// can be used:
+///
+/// ```rust
+/// const N: usize = 64;
+/// pub struct MyRngSeed(pub [u8; N]);
+///
+/// impl Default for MyRngSeed {
+///     fn default() -> MyRngSeed {
+///         MyRngSeed([0; N])
+///     }
+/// }
+///
+/// impl AsMut<[u8]> for MyRngSeed {
+///     fn as_mut(&mut self) -> &mut [u8] {
+///         &mut self.0
+///     }
+/// }
+/// ```
+///
+/// [`fill_bytes`]: Drbg::fill_bytes
+/// [`next_u32`]: Drbg::next_u32
+/// [`next_u64`]: Drbg::next_u64
+pub trait Drbg: Sized {
+    type Error: Debug;
```

Comment:
Is there a need for an associated error type? We control this trait, I'd rather just say what the error type *must* be for the internal subsystems (personally I'm more a fan of having distinct separate error types for different subsystems and then implementing `From` where it makes sense to provide sensible mapping, but I'm open to a general `FirmwareError` type for the time being if you disagree), and then the implementer has to provide the proper `From` mappings in some manner (via their trait impl or otherwise).

Reply:
So in this particular case we do control the trait. But we control it because no one else created it, for other cases I intend to use already existing traits from rustcrypto ecosystem like this: https://docs.rs/digest/latest/digest/trait.Digest.html


it is no final, but i wanted to have the flexibility that each hardware feature provider has its own error.